### ISSUE TITLE
Permutation Importance / Model template

### ISSUE THREAD

**Issue by lionelkusch** [mind-inria/hidimstat#131]
The actual implementation proposes a class for permutation importance, scikit-learn proposes a function and the PR #130 also proposes a function equivalent to the actual class.
I created similar tests in the PR #130 for comparing their usage.

The main difference between the function of sklearn and the PR#131 is that their options:
1. PR#131 supports permuting the variable by groups, which is not the case for sklearn
2. sklearn proposes to have to do multi-scoring/multi-loss at the same time, which is not possible in the PR#130
3. sklearn proposes to weight scoring for each sample, which is not possible in the PR#130
4. sklearn proposes to use only one part of the sample, which is not possible in the PR#130
5. The return of sklearn is a dictionary with an array: the mean and the std of importance for each score and the PR#130 returns only the means of importance.
6. sklearn only presents this method for random forest, and in their [MOOC](https://inria.github.io/scikit-learn-mooc/python_scripts/dev_features_importance.html], they suggest we implement the method ourselves.

I have 4 questions:
1. Do we use the implementation of scikit-learn for hidimstat?
2. Do we use the class or the function implementation for hidimstat?
3. Do we include all or a part of the options of scikit-learn function in hidimstat function/class, for having compatibility with scikit-learn?
4. Do we create a PR in scikit-learn for changing their function?

**Comment by lionelkusch** [comment]
@bthirion @AngelReyero @jpaillard

**Comment by jpaillard** [comment]
1. IMHO: 1. handling groups is a critical feature for high-dim problems; 2, 3, and 4: I don't see a strong drive for developing this feature, but I think that the current implementations should make it relatively easy to integrate such enhancements. 
2. As you described, the function is probably easier for inexperienced users, while classes offer more possibilities to expert users. Permutation importance is, however, probably not a good example. The possibilities of classes really whine for CPI where it allows conditional sampling or reusing a fitted importance model on various test data. Also, for LOCO where you can then score submodels $\mu_{(-j)} = \widehat{\mathbb{E}}[y|X^{(-j)}]$ that would not be available with a function. 
3. No strong opinion on this point
4. After exchanging with @glemaitre, the timeline and editorial process of scikit-learn is probably much slower than our current development. It's probably better to work independently on hidimstat.

**Comment by bthirion** [comment]
1. 2. 4. OK (Although there is a lot to gain for you and for sklearn in the long terms to have the discussion, so I really encourage you to collaborate on it; but indeed we have to go at a difference pace with hidimstat)
3. I'm not sure we should have FULL compatibility (my impression is that sklearn's implementation is a bit historical and  probably not grep form a usage perspective).

**Comment by bthirion** [comment]
I understand the pedagogical value, but I don't want to advertise permutation importance, which is plain wrong.
So very very very low priority.
Maybe we should have an example that shows that CPI should be done instead if one want p-values.

**Comment by lionelkusch** [comment]
I am not competent enough in statistics for evaluating some conceptual errors, but from a computational point of view, it is possible to apply the same methods to CPI for getting an estimation of p-values.

**Comment by bthirion** [comment]
Fair enough, but this actually the main point raised by @achamma723 in all his papers : the p-values returned by permutation importance are plain wrong as soon as there is a little bit of correlation btw variables. 
To me this is a case  where we could duplicate the example in 
https://mind-inria.github.io/hidimstat/auto_examples/plot_diabetes_variable_importance_example.html#sphx-glr-auto-examples-plot-diabetes-variable-importance-example-py
with simulated, null correlated data to show this behavior.

**Comment by jpaillard** [comment]
I tried to evidence this fact in [the classification example](https://mind-inria.github.io/hidimstat/auto_examples/plot_variable_importance_classif.html#sphx-glr-auto-examples-plot-variable-importance-classif-py) where it shows on simulated data that CPI identifies the support and PI makes false positives.
